File: 0022-generate-parentheses/0022-generate-parentheses.py
from collections import deque

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        answer = []

        def isValid(s):
            d = deque()
            for c in s:
                if c == "(":
                    d.append(c)
                    continue
                else:
                    if len(d) == 0: return False
                    opening = d.pop()
        
            if len(d) == 0:
                return True
            else:
                return False

        def dfs (p, opening, closing):
            if opening == closing == 0:
                if isValid(p):
                    answer.append(p)
                return

            if opening > 0:
                dfs(p+"(", opening-1, closing)
            if closing > 0:
                dfs(p+")", opening, closing-1)

        dfs("", n, n)
        return answer

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private List<String> answer = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        backtracking(n, n, "");
        return answer;
    }

    void backtracking(int open, int close, String s){
        if (open == 0 && close == 0){
            answer.add(s);
            return;
        }

        if (open > 0){
            backtracking(open-1, close, s + "(");
        } 
        if (close > open){
            backtracking(open, close-1, s + ")");
        }
    }
}